### DISCUSSION TITLE
Remove association requirement for entities that only atomically custody HTS tokens

### DISCUSSION THREAD

**Discussion by littletarzan** [hiero-ledger/hiero-improvement-proposals#696] [Ideas]
### Introduction

Smart contracts are required to be associated with all tokens they custody. This makes sense when the contract has a long lived balance > 0; Hedera state is expanded and must be paid for in ongoing maintenance and storage costs to the network. However, there are many use cases that call for smart contracts atomically taking custody of a token, often to transfer to another entity within the same contract call. In this case, association is still required but Hedera state is not expanded. We propose the idea and eventual HIP that association not be required for Hedera entities that only custody a token during an atomic transaction.

### Motivation

Smart contracts are often used to facilitate trustless transfer of tokens between parties, i.e. decentralized exchanges (DEXes). The DEXes on Hedera have different approaches to this. SaucerSwap uses delegatecall to avoid its router contract self-custodying tokens, HeliSwap optimistically associates tokens to its router contract for liquidity removal, and Pangolin's router associates and dissociates tokens to its router for liquidity removal. 

- The gas cost to associate tokens to contracts that only custody tokens atomically is not justified.
- The tech burden on avoiding association costs is high in the case of SaucerSwap. As of March 10, 2023, the precompile contract `0x167` cannot be delegatecalled.
- Ethereum feature parity is hampered by this association requirement, as many protocols on EVM networks allow self-custody of tokens by smart contracts without issue.

### Proposed Solution

Hedera Smart Contract Service optimistically allows custody of any token for any entity, and checks at the end of a contract call that tokens are associated to entities whose balance is changing from zero to nonzero.

**Comment by mgarbs** [comment]
Thanks for starting this discussion @littletarzan its a good one to have. I am curious to see what the rest of the community thinks about this. My take-

Pros:
Reduced gas costs: By removing the association requirement of contract intermediaries in atomic transactions, the overall gas of the transaction is reduced.

Enhanced interoperability: I see both sides here. This proposal brings Hedera closer to Ethereum feature parity, but because we are different networks, it might not always be the best approach to achieve exact parity.

Encourages trustless transfers: By removing the association requirement for internal transactions with net zero balances, smart contracts can facilitate trustless transfers of tokens between parties more efficiently, ultimately promoting growth.

Cons:
Potential for increased complexity: Optimistically allowing custody of any token for any entity might increase the complexity of the Hedera Smart Contract Service, potentially leading to unforeseen issues or vulnerabilities.

Maintenance overhead: Implementing this proposal would require modifications to the existing Hedera Smart Contract Service, which could introduce additional maintenance overhead for Engineering.

Potential abuse: There might be a risk of malicious actors taking advantage of the relaxed association requirements to manipulate token custody, although this could be mitigated by ensuring proper checks at the end of a contract call. So this part would need very careful consideration

So tldr: Its cool but could be dangerous so carefully thinking through the implementation is crucial. @Nana-EC what do you think about this proposal?

**Comment by kpachhai** [comment]
@littletarzan Since it has been a while since this discussion was created, wondering if there's a need to push this forward or perhaps withdraw this for now? If there's a need for it, we can continue the discussions further. If not, feel free to close the issue.

**Comment by littletarzan** [comment]
@kpachhai Let's withdraw this idea
